src/lib/esnext.promise.d.ts

```diff
@@ -0,0 +1,14 @@
+/**
+ * Represents the completion of an asynchronous operation
+ */
+interface Promise<T> {
+    /**
+     * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
+     * resolved value cannot be modified from the callback.
+     * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
+     * @returns A Promise for the completion of the callback.
+     */
+    finally<TResult = never>(
+        onfinally?: (() => TResult | PromiseLike<TResult>) | undefined | null
+    ): Promise<T | TResult>;

```

Comment:
The result of the `onfinally` callback is **always ignored** (unless it's a rejection). The way it works on the spec is pretty complicated, but it essentially wraps the `onfinally` argument with a function that returns a Promise with the original Promise's state.

```ts
finally(onfinally?: (() => void) | undefined | null): Promise<T>
```
is the correct signature, unless we want to provide some way to support the result of `Promise.reject` as a return value; but that can be done just as well by `throw`ing inside the `onfinally`.

Reply:
@Kovensky thanks for the review.
I have one remark though, going with the return type also means that the compiler won't complain if onfinally simply throws an error or returns any thing (since that is assignable to void). I guess that this is what we would like to have.
Also, it is indeed possible to have a Promise.reject (assignable to void) within finally
```typescript
const generatePromise = (value)=> new Promise((resolve,reject)=>{
	resolve(value);
});
generatePromise(13)
  .finally(() => Promise.reject(10))
  .then(data=>console.log(data))
  .catch(data=>console.log(data))
```
I see however, that  [DefinitelyTyped definition for promise.prototype.finally](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/promise.prototype.finally/index.d.ts) is not going in the same direction as you proposed
```typescript
    interface Promise<T> {
        finally<U>(onFinally?: () => U | PromiseLike<U>): Promise<T>;
    }
```
I would like to have a confirmation first before making the change.
@andy-ms I see that at some point you contributed to the file containing [the definitions for promise.prototype.finally](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/promise.prototype.finally/index.d.ts) , could you please give us a recommendation on which definition to adopt

Thanks!

Reply:
I think returning `void` is best. Accepting `U` for an arbitrary value of `U` just means accepting anything after all.
(I didn't contribute that definition; I just moved the file.)

Reply:
Thanks @andy-ms @Kovensky 
PR updated